00_Base/src/datasources/TokensDatasource.ts

```diff
@@ -0,0 +1,229 @@
+    }
+    //
+    // const newOCPPAuth =
+    //   OCPITokensMapper.mapOcpiTokenToOcppAuthorization(updatedToken);
+    // newOCPPAuth.id = existingOCPIToken.id;
+    // const updatedOcppAuth =
+    //   await this.authorizationRepository.createOrUpdateByQuerystring(
+    //     newOCPPAuth,
+    //     {
+    //       idToken: partialToken.uid!,
+    //       type: OCPITokensMapper.mapOcpiTokenTypeToOcppIdTokenType(partialToken.type),
+    //     },
+    //   );
+    //
+    // if (!updatedOcppAuth) {
+    //   throw new Error('Could not save token');
+    // }
+
+    return new TokenDTO();
+  }
+
+  async updateToken(tokenDto: TokenDTO): Promise<TokenDTO> {
+    const mappedOcppAuth =
+      OcpiTokensMapper.mapOcpiTokenToOcppAuthorization(tokenDto);
+    const savedOcppAuth =
+      await this.authorizationRepository.createOrUpdateByQuerystring(
+        mappedOcppAuth,
+        {
+          idToken: tokenDto.uid,
+          type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+            tokenDto.type,
+          ),
+        },
+      );
+
+    if (!savedOcppAuth) {
+      throw new UnknownTokenException(
+        'Authorization could not be created upon saving token',
+      );
+    }
+
+    try {
+      const ocpiToken = await this.tokensRepository.createOrUpdateOcpiToken(
+        savedOcppAuth?.id,
+        tokenDto,
+      );
+      return OcpiTokensMapper.toDto(savedOcppAuth!, ocpiToken);
+    } catch (error) {
+      this.logger.error('Error saving token', error);
+      throw error;
+    }
+  }
+
+  async updateBatchedTokens(tokens: TokenDTO[]): Promise<void> {
+    const batchFailedTokens: TokenDTO[] = [];
+
+    // TODO: do an actual bulk update
+    for (const token of tokens) {
+      try {
+        await this.updateToken(token);
+      } catch (e) {
+        this.logger.error(e);
+        batchFailedTokens.push(token);
+      }
+    }
+
+    if (batchFailedTokens.length > 0) {
+      throw new Error(
+        `Failed to fetch tokens: ${JSON.stringify(batchFailedTokens)}`,
+      );
+    }
+  }
+
+  async getTokenDtoByIdToken(
+    idToken: string,
+    type: IdTokenEnumType,
+  ): Promise<TokenDTO | undefined> {
+    const ocppAuth =
+      await this.authorizationRepository.readOnlyOneByQuerystring({
+        idToken,
+        type,
+      });
+
+    if (!ocppAuth) {
+      // TODO better error
+      return undefined;
+    }
+
+    const ocpiToken = await this.getOcpiTokenByAuthorizationId(ocppAuth.id);
+
+    if (!ocpiToken) {
+      // TODO better error
+      return undefined;
+    }
+
+    return OcpiTokensMapper.toDto(ocppAuth, ocpiToken);
+  }
+
+  async getOcpiTokenByAuthorizationId(
+    authorizationId: string,
+  ): Promise<OcpiToken | undefined> {
+    const ocpiToken = await this.tokensRepository.readOnlyOneByQuery({
+      where: {
+        authorization_id: authorizationId,
+      },
+    });
+    if (!ocpiToken) {
+      // TODO better error
```

Comment:
Can we fill out this todo now?

Reply:
Lmk what you think but I kinda like returning undefined here instead of throwing.

I think throwing makes sense when you are trying to patch a token that does not exist.

Reply:
I agree, get is ok to return undefined, patch can throw as the operation is erroring. So let's remove the `TODO` here :)